woocommerce/woocommerce-ios
# [Reliability] Add fallback for decoding variation objects for Product variations

Closes: #9841, #5512


## Description

This PR increases the app's resilience when the app receives an array of variation objects instead of an array of variation IDs (the expected value type) for the `variations` field on `Product`. (This is a known decoding error that users have experienced with certain third-party plugins.)

We had a fallback to return an empty variations list in case of errors like this (so the product list parsing didn't fail entirely), but this PR adds support for parsing the response so we can display the variations in variable product details.

If the expected decoding for the `variations` field fails, we try decoding the variation IDs from variation objects. We don't decode these as `[ProductVariation]` because that requires `decoder.userInfo` to contain the `productID` (which we won't have if we're decoding a product list). Instead, we try keying each object using the `ProductVariation` coding keys and then decode the variation ID from the `id` field.

## Testing instructions


1. Create at least one variable product with variations on your store.
2. Install and activate the [wePOS plugin](https://wordpress.org/plugins/wepos/). (This is one plugin known to change the API response like this.)
3. Build and run the app.
4. Open the Products tab and confirm the product list loads.
5. Select a variable product and confirm the product details show the expected variations count.
6. Open the Variations section and confirm the variation list loads with the expected variations (not empty).

## Screenshots

Before|After
-|-
![Simulator Screen Shot - iPhone 14 Pro - 2023-06-02 at 11 39 52](https://github.com/woocommerce/woocommerce-ios/assets/8658164/6fb540d5-5602-4a64-ad33-1b6e64376f19)|![Simulator Screen Shot - iPhone 14 Pro - 2023-06-02 at 11 45 56](https://github.com/woocommerce/woocommerce-ios/assets/8658164/2fb821e7-5c2a-43e3-ace2-48864e163b36)
![Simulator Screen Shot - iPhone 14 Pro - 2023-06-02 at 11 39 56](https://github.com/woocommerce/woocommerce-ios/assets/8658164/1b0e03d9-d1ba-492b-a5f4-9465d796e3e7)|![Simulator Screen Shot - iPhone 14 Pro - 2023-06-02 at 11 46 00](https://github.com/woocommerce/woocommerce-ios/assets/8658164/e84e3543-86a0-4c53-985a-788f48595eba)

---
